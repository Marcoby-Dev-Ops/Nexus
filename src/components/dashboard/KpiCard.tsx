/**
 * @name KpiCard
 * @description KPI summary card for displaying a metric.
 * @param {KpiCardProps} props
 * @returns {JSX.Element}
 */
import PropTypes from 'prop-types';
type KpiCardProps = { title: string; value: string | number; delta?: string };

export function KpiCard({ title, value, delta }: KpiCardProps) {
  return (
    <div className="rounded-xl border p-4 shadow-sm">
      <p className="text-muted-foreground mb-1 text-xs uppercase">{title}</p>
      <p className="text-3xl font-semibold">{value}</p>
      {delta && <p className="text-xs text-green-600">{delta}</p>}
    </div>
  );
}

KpiCard.propTypes = {
  title: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  delta: PropTypes.string,
}; 