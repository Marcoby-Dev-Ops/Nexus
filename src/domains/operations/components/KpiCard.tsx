import React from 'react';
import { LineChart, Line, ResponsiveContainer } from 'recharts';
import { Card } from '@/components/ui/Card';
import type { KPI } from '../types';
import { useRunPlaybook } from '../hooks/useRunPlaybook';
import { Button } from '@/components/ui/Button';

interface Props {
  kpi: KPI;
  isFocus?: boolean; // if this KPI is the worst contributor
  advice?: string | string[];
}

function formatDelta(delta?: number) {
  if (delta === undefined) return '—';
  const sign = delta > 0 ? '+' : '';
  return `${sign}${delta.toFixed(1)}%`;
}

export const KpiCard: React.FC<Props> = ({ kpi, isFocus, advice }) => {
  const deltaColor = kpi.delta && kpi.delta >= 0 ? 'text-green-600' : 'text-destructive';

  const { mutate: runPlaybook, isPending } = useRunPlaybook();

  return (
    <Card className={isFocus ? 'border-2 border-destructive' : ''}>
      <div className="p-4 space-y-2">
        <div className="flex justify-between items-center">
          <h4 className="font-medium text-sm">{kpi.label}</h4>
          <span className={`text-xs ${deltaColor}`}>{formatDelta(kpi.delta)}</span>
        </div>
        <div className="text-2xl font-semibold">{kpi.value}</div>
        {kpi.history && (
          <div className="h-10 w-full">
            <ResponsiveContainer width="100%" height="100%">
              <LineChart data={kpi.history.map((v, i) => ({ idx: i, v }))} margin={{ top: 2, bottom: 2, left: 0, right: 0 }}>
                <Line type="monotone" dataKey="v" stroke="#0891b2" strokeWidth={2} dot={false} />
              </LineChart>
            </ResponsiveContainer>
          </div>
        )}
        {isFocus && advice && (
          <div className="pt-2 text-sm border-t border-border space-y-2">
            <p className="font-medium">AI Recommendation</p>
            <p>{advice}</p>
            <Button
              isLoading={isPending}
              disabled={isPending}
              onClick={() => runPlaybook(kpi.id)}
              variant="outline"
              size="sm"
            >
              {isPending ? 'Queuing…' : 'Run Playbook'}
            </Button>
          </div>
        )}
      </div>
    </Card>
  );
};

export default KpiCard; 