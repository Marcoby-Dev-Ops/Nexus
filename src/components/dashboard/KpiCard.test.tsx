/**
 * @file KpiCard.test.tsx
 * @description Unit and snapshot tests for the KpiCard dashboard component.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { KpiCard } from './KpiCard';

describe('KpiCard', () => {
  it('renders title and value', () => {
    render(<KpiCard title="Revenue" value="$1000" />);
    expect(screen.getByText('Revenue')).toBeInTheDocument();
    expect(screen.getByText('$1000')).toBeInTheDocument();
  });

  it('renders delta with correct color (positive)', () => {
    render(<KpiCard title="Revenue" value="$1000" delta="+10%" />);
    const delta = screen.getByText('+10%');
    expect(delta).toHaveClass('text-green-600');
  });

  it('renders delta with correct color (negative)', () => {
    render(<KpiCard title="Revenue" value="$1000" delta="-5%" />);
    const delta = screen.getByText('-5%');
    expect(delta).toHaveClass('text-red-600');
  });

  it('renders sparkline if sparklineData is provided', () => {
    render(<KpiCard title="Revenue" value="$1000" sparklineData={[1,2,3,4,5]} />);
    expect(screen.getByRole('img', { hidden: true })).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<KpiCard title="Revenue" value="$1000" delta="+10%" sparklineData={[1,2,3,4,5]} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 