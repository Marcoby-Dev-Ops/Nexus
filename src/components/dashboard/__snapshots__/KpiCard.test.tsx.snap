// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KpiCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-xl border p-4 shadow-sm flex flex-col gap-1"
  >
    <p
      class="text-muted-foreground mb-1 text-xs uppercase"
    >
      Revenue
    </p>
    <div
      class="flex items-end justify-between gap-2"
    >
      <p
        class="text-3xl font-semibold"
      >
        $1000
      </p>
      <svg
        class="text-primary/70"
        fill="none"
        height="18"
        viewBox="0 0 64 18"
        width="64"
      >
        <polyline
          fill="none"
          points="0,16 15,13 30,10 45,7 60,4"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="text-xs font-medium text-success"
    >
      +10%
    </p>
  </div>
</DocumentFragment>
`;
